Add import path support and fix binary encoding for k6/JavaScript interop

brianV opened the pull request:
### Summary

Adds support for proto file imports and fixes binary data handling to enable real-world usage in k6 load testing. These changes allow xk6-protobuf to work with complex proto definitions that use Google Well-Known Types and nested imports, while ensuring encoded messages remain binary-safe when passed through JavaScript.

### Problem

When using xk6-protobuf in k6 load tests with realistic proto definitions, we encountered three blocking issues:

#### 1. Import Resolution Failure
Proto files using imports failed to compile:
```
open google/protobuf/duration.proto: no such file or directory
```

Example proto that would fail:
```protobuf
syntax = "proto3";
package iot.test_messages;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";

message Ping {
  int64 id = 1;
  google.protobuf.Timestamp sent_time = 2;
}
```

**Root cause**: The `SourceResolver` had no `ImportPaths` configured, so `protocompile` couldn't resolve any imports.

#### 2. Path Concatenation Issue
After adding ImportPaths, loading still failed with doubled paths:
```
open /path/to/proto//path/to/proto/messages.proto: no such file or directory
```

**Root cause**: `protocompile` treats the proto file path as relative to ImportPaths and concatenates them. We needed to convert absolute paths to relative paths.

#### 3. Binary Data Corruption in JavaScript
Encoded protobuf messages were corrupted when used in k6 scripts. Comparing hex dumps of the same message showed UTF-8 double-encoding:
```
Expected: 0a 0f 74 65 73 74 2d 64 65 76 69 63 65  (field 1, string)
Actual:   c2 9b c3 b3 c2 bf c2 9b c3 87...         (UTF-8 encoded bytes)
```

**Root cause**: When `Encode()` returned `[]byte`, k6's JavaScript environment converted it using `String.fromCharCode()`. This causes bytes > 127 to become multi-byte UTF-8 sequences:
- Byte `0x9b` → Unicode U+009B → UTF-8 `c2 9b` (2 bytes)
- Byte `0xb3` → Unicode U+00B3 → UTF-8 `c2 b3` (2 bytes)

This corrupts the protobuf wire format, making messages undecodable by downstream services.

### Solution

#### 1. ImportPaths Support
```go
compiler := protocompile.Compiler{
    Resolver: &protocompile.SourceResolver{
        ImportPaths: importPaths,  // Enable proto import resolution
    },
}
```

Defaults to the proto file's directory if no import paths provided, enabling automatic resolution of relative imports.

#### 2. Relative Path Conversion
```go
absProtoFile, _ := filepath.Abs(protoFilePath)
relPath, _ := filepath.Rel(importPaths[0], absProtoFile)
protoFilePath = relPath  // Use relative path for compilation
```

Converts absolute paths to paths relative to the import directory, preventing path concatenation issues.

#### 3. Binary-Safe String Return
```go
func (p *ProtoFile) Encode(jsonString string) string {
    // ... marshal to []byte ...
    return string(data)  // Go string preserves all bytes 0-255
}
```

Changed return type from `[]byte` to `string`. In Go, strings are binary-safe and preserve all byte values 0-255. When k6 receives this string from Go, it preserves the raw bytes without UTF-8 re-encoding, allowing binary data to be used correctly (e.g., published to message brokers, written to files, etc.).

#### 4. Fully Qualified Name Support
Added logic to extract simple names from fully qualified message names (e.g., `iot.test_messages.Ping` → `Ping`) for compatibility with `ByName()` lookups.

### Testing

Validated with a k6 load test that:
- ✅ Loads proto files with Google WKT imports (`google.protobuf.Timestamp`, `google.protobuf.Duration`)
- ✅ Loads proto files with nested package imports
- ✅ Encodes messages that decode correctly in downstream Go services (wire-compatible)
- ✅ Binary data passes through JavaScript without corruption

**Before**: Services reported `proto: cannot parse invalid wire-format data`  
**After**: Messages decode successfully with all fields intact

### Use Case

This enables using xk6-protobuf for realistic load testing scenarios where:
- Proto definitions use imports (standard practice in production)
- Encoded messages need to be sent over the network or processed by other k6 extensions
- Downstream services need to decode the messages (requires wire-compatible protobuf)

### Backward Compatibility

✅ **Fully backward compatible**
- `importPaths` parameter is optional (defaults to proto file directory)
- Return type change from `[]byte` to `string` is transparent to JavaScript consumers
- Existing code without imports continues to work

brianV committed:
Enhance Load function to accept import paths

Added support for custom import paths in Load function.

brianV committed:
Update import paths for proto file loading

brianV committed:
Fix importPaths initialization in Load function

brianV committed:
Make import paths absolute in Load function

Convert import path to absolute if possible.

brianV committed:
Enhance logging for proto path resolution

Added debug logging for proto directory and file path resolution.

brianV committed:
Change proto file path to relative from import path

brianV committed:
Refactor Load function for improved clarity

brianV committed:
Fix message descriptor retrieval in protobuf.go

brianV committed:
Improve logging in Encode method

Enhanced logging in Encode method for better debugging.

brianV committed:
Enhance logging in Encode method of ProtoFile

Updated logging in Encode method to preview JSON input and count fields set.

brianV committed:
Change Encode method to return string

Refactor Encode method to return a string instead of byte slice and remove debug logs.

brianV committed:
Refactor protobuf loading and encoding functions

brianV committed:
Clean up: remove unnecessary changes

Keep only the essential changes:
- Add importPaths parameter and ImportPaths support
- Add relative path conversion
- Add fully qualified name support
- Change Encode() return type to string

Remove unnecessary changes:
- Import reordering
- Variable renames
- Error handling changes (log.Fatal to panic)
- All changes to Decode() function
- Comments and cosmetic changes

brianV committed:
Convert relative import paths to absolute paths

When import paths are provided as relative paths, they need to be
converted to absolute paths before passing to protocompile. This
ensures that filepath.Rel() can correctly compute the relative path
between the proto file and import path.

Fixes issue where protobuf files with imports could not be loaded
when using relative import paths.

JBartscher commented:
@brianV any updates on this? I face the exact same issue and I like your fix! Would be highly appreciated :relaxed:

brianV commented:
> @brianV any updates on this? I face the exact same issue and I like your fix! Would be highly appreciated ☺️

At this point, t's up to the project maintainers to merge it or further refine it. I used the code to accomplish what I needed, then contributed it back upstream.

JBartscher commented:
@gusev-klara ?

Donatas-L commented:
I have requested @fildawtraveltime to review this PR

fildawtraveltime committed:
Fix default import paths, path resolution, and FQN extraction

  - Default import paths to CWD + proto file directory instead of only
    the proto file's directory. Using only the proto directory broke
    imports with paths relative to the project root.
  - Resolve proto file path against all import paths, not just the first.
    Fail fast if the file isn't under any of them.
  - Use strings.LastIndex for FQN extraction and guard against empty
    message names.
  - Replace silent error fallbacks in filepath.Abs/Rel with log.Fatalf.
  - Add tests.

fildawtraveltime commented on protobuf.go:
```diff
@@ -23,9 +24,41 @@ type ProtoFile struct {
+		if err != nil {
+			absProtoDir = protoDir
+		}
+		importPaths = []string{absProtoDir}
+	} else {
+		// Convert all provided import paths to absolute paths
+		absImportPaths := make([]string, len(importPaths))
+		for i, path := range importPaths {
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				absPath = path
+			}
+			absImportPaths[i] = absPath
+		}
+		importPaths = absImportPaths
+	}
+
 	compiler := protocompile.Compiler{
-		Resolver: &protocompile.SourceResolver{},
+		Resolver: &protocompile.SourceResolver{
+			ImportPaths: importPaths,
+		},
+	}
+
+	// Make proto file path relative to import path for protocompile
+	absProtoFile, err := filepath.Abs(protoFilePath)
+	if err == nil {
+		relPath, err := filepath.Rel(importPaths[0], absProtoFile)
+		if err == nil {
+			protoFilePath = relPath
+		}
 	}
```
Addressed - all filepath.Abs calls now log.Fatalf on failure, consistent with the rest of the file.

fildawtraveltime commented on protobuf.go:
```diff
@@ -23,9 +24,41 @@ type ProtoFile struct {
+		if err == nil {
+			protoFilePath = relPath
+		}
 	}
```
Addressed - the relativization now iterates all import paths and picks the first one that contains the proto file. Also added a guard that fails if the file isn't under any of the provided import paths.

fildawtraveltime commented on protobuf.go:
```diff
@@ -39,10 +72,24 @@ func (p *Protobuf) Load(protoFilePath, lookupType string) ProtoFile {
+			simpleName = lookupType[i+1:]
+			break
+		}
+	}
```
Addressed - added a guard for this case

fildawtraveltime commented on protobuf.go:
```diff
@@ -39,10 +72,24 @@ func (p *Protobuf) Load(protoFilePath, lookupType string) ProtoFile {
 	}
 
-	return ProtoFile{files[0].Messages().ByName(protoreflect.Name(lookupType))}
+	// Extract simple name from fully qualified name (e.g., "iot.test_messages.Ping" -> "Ping")
+	simpleName := lookupType
+	for i := len(lookupType) - 1; i >= 0; i-- {
+		if lookupType[i] == '.' {
+			simpleName = lookupType[i+1:]
+			break
+		}
+	}
+
+	messageDesc := files[0].Messages().ByName(protoreflect.Name(simpleName))
+	if messageDesc == nil {
+		log.Fatalf("Message type %s not found", lookupType)
+	}
```
Already covered by the empty simpleName guard.

fildawtraveltime commented on protobuf.go:
```diff
@@ -23,9 +24,41 @@ type ProtoFile struct {
 }
 
-func (p *Protobuf) Load(protoFilePath, lookupType string) ProtoFile {
+func (p *Protobuf) Load(protoFilePath, lookupType string, importPaths ...string) ProtoFile {
```
Not needed — compiler.Compile will fail with a clear error. An os.Stat pre-check would duplicate that.

fildawtraveltime commented on protobuf.go:
```diff
@@ -39,10 +72,24 @@ func (p *Protobuf) Load(protoFilePath, lookupType string) ProtoFile {
 	}
 
-	return ProtoFile{files[0].Messages().ByName(protoreflect.Name(lookupType))}
+	// Extract simple name from fully qualified name (e.g., "iot.test_messages.Ping" -> "Ping")
+	simpleName := lookupType
+	for i := len(lookupType) - 1; i >= 0; i-- {
+		if lookupType[i] == '.' {
+			simpleName = lookupType[i+1:]
+			break
+		}
+	}
+
+	messageDesc := files[0].Messages().ByName(protoreflect.Name(simpleName))
+	if messageDesc == nil {
+		log.Fatalf("Message type %s not found", lookupType)
+	}
+
+	return ProtoFile{messageDesc}
```
Not a regression. The old code passed the full FQN to ByName(), which never matched since message names don't include the package prefix. Nested message resolution is a pre-existing limitation of files[0].Messages().

gergelytraveltime commented:
/review
